src/v/security/gssapi_authenticator.cc

```diff
@@ -0,0 +1,514 @@
+    friend std::ostream& operator<<(std::ostream& os, hex const& h) {
+        for (size_t i = 0; i < h._v.length(); ++i) {
+            if ((i % 16) == 0) {
+                os << '\n';
+            }
+            os << std::setfill('0') << std::setw(2) << std::hex
+               << (int)(h._v[i]) << ' ';
+        }
+        return os;
+    }
+    bytes_view _v;
+};
+
+/*
+ * Function: display_status
+ *
+ * Purpose: displays GSS-API messages
+ *
+ * Arguments:
+ *
+ *      msg             a string to be displayed with the message
+ *      maj_stat        the GSS-API major status code
+ *      min_stat        the GSS-API minor status code
+ *
+ * Effects:
+ *
+ * The GSS-API messages associated with maj_stat and min_stat are
+ * displayed on stderr, each preceeded by "GSS-API error <msg>:
+" and
+ * followed by a newline.
+ */
+void display_status(const char* msg, OM_uint32 maj_stat, OM_uint32 min_stat) {
+    display_status_1(msg, maj_stat, GSS_C_GSS_CODE);
+    display_status_1(msg, min_stat, GSS_C_MECH_CODE);
+}
+
+ss::future<result<bytes>>
+gssapi_authenticator::authenticate(bytes_view auth_bytes) {
+    vlog(seclog.info, "gss {} authenticate", _state);
+    switch (_state) {
+    case state::init: {
+        auto principal = config::shard_local_cfg().sasl_kerberos_principal();
+        auto keytab = config::shard_local_cfg().sasl_kerberos_keytab();
+
+        if (auto res = co_await _worker.submit(
+              [this, principal, keytab]() { return init(principal, keytab); });
+            res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+            co_return res.assume_error();
+        }
+    }
+        [[fallthrough]];
+    case state::more: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return more(auth_bytes); });
+        if (res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+        }
+        co_return res;
+    }
+    case state::ssfcap: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return ssfcap(auth_bytes); });
+        if (res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+        }
+        co_return res;
+    }
+    case state::ssfreq: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return ssfreq(auth_bytes); });
+        if (res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+        }
+        co_return res;
+    }
+    case state::complete:
+    case state::failed:
+        break;
+    }
+    vlog(seclog.info, "gss {} authenticate failed", _state);
+    _state = state::failed;
+    co_return errc::invalid_scram_state;
+}
+
+result<void>
+gssapi_authenticator::init(ss::sstring principal, ss::sstring keytab) {
```

Comment:
Should there be an assert check or something to ensure that when `init` is called we are in the `init` state?